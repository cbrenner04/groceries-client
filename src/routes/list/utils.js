import { toast } from 'react-toastify';

import axios from '../../utils/api';
import { formatDueBy } from '../../utils/format';

export function itemName(item, listType) {
  return {
    BookList: `${item.title ? `"${item.title}"` : ''} ${item.author || ''}`,
    GroceryList: `${item.quantity || ''} ${item.product || ''}`,
    MusicList:
      `${item.title ? `"${item.title}"` : ''} ${item.artist || ''}` +
      `${item.artist && item.album ? ' - ' : ''}${item.album || ''}`,
    ToDoList: `${item.task}`,
  }[listType].trim();
}

export function mapIncludedCategories(items) {
  const cats = [''];
  items.forEach((item) => {
    if (!item.category) {
      return;
    }
    const cat = item.category.toLowerCase();
    if (!cats.includes(cat)) {
      cats.push(cat);
    }
  });
  return cats;
}

export function categorizeNotPurchasedItems(items, categories) {
  const obj = {};
  categories.forEach((cat) => {
    obj[cat] = [];
  });
  items.forEach((item) => {
    if (!item.category) {
      obj[''].push(item);
      return;
    }
    const cat = item.category.toLowerCase();
    if (!obj[cat]) {
      obj[cat] = [];
    }
    obj[cat].push(item);
  });
  return obj;
}

function performSort(items, sortAttrs) {
  // return when all items are sorted
  if (sortAttrs.length === 0) {
    return items;
  }
  const sortAttr = sortAttrs.pop();
  const sorted = items.sort((a, b) => {
    // the sort from the server comes back with items with number_in_series: `null` at the end of the list
    // without the next two lines this would put those items at the front of the list
    if (a[sortAttr] === null) {
      return 1;
    }
    if (b[sortAttr] === null) {
      return -1;
    }
    const positiveBranch = a[sortAttr] > b[sortAttr] ? 1 : 0;
    return a[sortAttr] < b[sortAttr] ? -1 : positiveBranch;
  });
  return performSort(sorted, sortAttrs);
}

export function sortItems(listType, items) {
  let sortAttrs = [];
  if (listType === 'BookList') {
    sortAttrs = ['author', 'number_in_series', 'title'];
  } else if (listType === 'GroceryList') {
    sortAttrs = ['product'];
  } else if (listType === 'MusicList') {
    sortAttrs = ['artist', 'album', 'title'];
  } else if (listType === 'ToDoList') {
    sortAttrs = ['due_by', 'assignee_id', 'task'];
  }
  const sorted = performSort(items, sortAttrs);
  return sorted;
}

export async function fetchList({ id, history }) {
  try {
    // TODO: looks like list.users_list_id is not coming back
    const responses = await Promise.all([axios.get(`/lists/${id}/users_lists`), axios.get(`/lists/${id}`)]);
    const [
      {
        data: { accepted, pending },
      },
      {
        data: {
          current_user_id: currentUserId,
          not_purchased_items: responseNotPurchasedItems,
          purchased_items: purchasedItems,
          list,
          categories,
        },
      },
    ] = responses;
    const userInAccepted = accepted.find((acceptedList) => acceptedList.user.id === currentUserId);
    if (!userInAccepted) {
      toast('List not found', { type: 'error' });
      history.push('/lists');
      return;
    }
    const allAcceptedUsers = accepted.map(({ user }) => user);
    const allPendingUsers = pending.map(({ user }) => user);
    const listUsers = allAcceptedUsers.concat(allPendingUsers);
    const includedCategories = mapIncludedCategories(responseNotPurchasedItems);
    const notPurchasedItems = categorizeNotPurchasedItems(responseNotPurchasedItems, includedCategories);
    const { permissions } = userInAccepted.users_list;

    return {
      currentUserId,
      list,
      purchasedItems,
      categories,
      listUsers,
      includedCategories,
      notPurchasedItems,
      permissions,
    };
  } catch ({ response }) {
    if (response) {
      if (response.status === 401) {
        toast('You must sign in', { type: 'error' });
        history.push('/users/sign_in');
        return;
      } else if ([403, 404].includes(response.status)) {
        toast('List not found', { type: 'error' });
        history.push('/lists');
        return;
      }
    }
    // any other errors we will catch and render generic UnknownError
    throw new Error();
  }
}

export async function fetchListToEdit({ id, history }) {
  try {
    const {
      data: {
        list: { owner_id, id: listId, name, completed, type },
        current_user_id: currentUserId,
      },
    } = await axios.get(`/lists/${id}/edit`);
    if (owner_id !== currentUserId) {
      toast('List not found', { type: 'error' });
      history.push('/lists');
      return;
    }
    return {
      listId,
      name,
      completed,
      type,
    };
  } catch ({ response }) {
    if (response) {
      if (response.status === 401) {
        toast('You must sign in', { type: 'error' });
        history.push('/users/sign_in');
        return;
      } else if ([403, 404].includes(response.status)) {
        toast('List not found', { type: 'error' });
        history.push('/lists');
        return;
      }
    }
    // any other errors will just be caught and render the generic UnknownError
    throw new Error();
  }
}

export async function fetchItemToEdit({ itemId, listId, itemType, history }) {
  try {
    // TODO: do i need all of this?
    // TODO: confirm required items always come back
    const [editListResponse, usersListsResponse, listResponse] = await Promise.all([
      axios.get(`/lists/${listId}/${itemType}/${itemId}/edit`),
      axios.get(`/lists/${listId}/users_lists`),
      axios.get(`/lists/${listId}`),
    ]);
    const {
      data: {
        item,
        list: { id: returnedListId, type },
      },
    } = editListResponse;
    const {
      data: { accepted, pending, current_user_id: currentUserId },
    } = usersListsResponse;
    const {
      data: { categories },
    } = listResponse;
    const userId = item.user_id;
    const returnedItemId = item.id;
    const product = item.product || '';
    const task = item.task || '';
    const purchased = item.purchased || false;
    const quantity = item.quantity || '';
    const completed = item.completed || false;
    const author = item.author || '';
    const title = item.title || '';
    const read = item.read || false;
    const artist = item.artist || '';
    const album = item.album || '';
    const dueBy = formatDueBy(item.due_by); // TODO: this will default to today. ok?
    const acceptedUsers = accepted.map(({ user }) => user);
    const pendingUsers = pending.map(({ user }) => user);
    const listUsers = acceptedUsers.concat(pendingUsers);
    const userInAccepted = accepted.find((acceptedList) => acceptedList.user.id === currentUserId);
    const assigneeId = item.assignee_id ? String(item.assignee_id) : ''; // TODO: really?
    const numberInSeries = item.number_in_series ? Number(item.number_in_series) : 0;
    const category = item.category || '';
    if (!userInAccepted || userInAccepted.users_list.permissions !== 'write') {
      toast('Item not found', { type: 'error' });
      history.push(`/lists/${listId}`);
      return;
    }
    return {
      listUsers,
      userId,
      list: {
        id: returnedListId,
        type,
        categories,
      },
      item: {
        id: returnedItemId,
        product,
        task,
        purchased,
        quantity,
        completed,
        author,
        title,
        read,
        artist,
        dueBy,
        assigneeId,
        album,
        numberInSeries,
        category,
      },
    };
  } catch ({ response }) {
    if (response) {
      if (response.status === 401) {
        toast('You must sign in', { type: 'error' });
        history.push('/users/sign_in');
        return;
      } else if ([403, 404].includes(response.status)) {
        toast('Item not found', { type: 'error' });
        history.push(`/lists/${listId}`);
        return;
      }
    }
    // any other errors will just be caught and render the generic UnknownError
    throw new Error();
  }
}
