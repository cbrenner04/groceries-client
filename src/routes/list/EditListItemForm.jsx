import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { Button, Form } from 'react-bootstrap';

import { defaultDueBy, formatDueBy, listTypeToSnakeCase } from '../../utils/format';
import { setUserInfo } from '../../utils/auth';
import Alert from '../../components/Alert';
import BookListItemFormFields from './components/BookListItemFormFields';
import GroceryListItemFormFields from './components/GroceryListItemFormFields';
import MusicListItemFormFields from './components/MusicListItemFormFields';
import ToDoListItemFormFields from './components/ToDoListItemFormFields';
import axios from '../../utils/api';

function EditListItemForm(props) {
  const [userId, setUserId] = useState(0);
  const [listId, setListId] = useState(0);
  const [itemId, setItemId] = useState(0);
  const [listType, setListType] = useState('GroceryList');
  const [listUsers, setListUsers] = useState([]);
  const [categories, setCategories] = useState([]);
  const [errors, setErrors] = useState('');
  const [product, setProduct] = useState('');
  const [task, setTask] = useState('');
  const [purchased, setPurchased] = useState(false);
  const [quantity, setQuantity] = useState('');
  const [completed, setCompleted] = useState(false);
  const [author, setAuthor] = useState('');
  const [title, setTitle] = useState('');
  const [read, setRead] = useState(false);
  const [artist, setArtist] = useState('');
  const [dueBy, setDueBy] = useState(defaultDueBy());
  const [assigneeId, setAssigneeId] = useState('');
  const [album, setAlbum] = useState('');
  const [numberInSeries, setNumberInSeries] = useState(0);
  const [category, setCategory] = useState('');

  useEffect(() => {
    if (props.match) {
      const { id, list_id } = props.match.params;
      const failure = ({ response, request, message }) => {
        if (response) {
          setUserInfo(response.headers);
          if (response.status === 401) {
            // TODO: how do we pass error messages along?
            props.history.push('/users/sign_in');
          } else {
            // TODO: how do we pass error messages along?
            props.history.push(`/lists/${list_id}`);
          }
        } else if (request) {
          // TODO: what do here?
        } else {
          setErrors(message);
        }
      };
      const headers = JSON.parse(sessionStorage.getItem('user'));
      Promise.all([
        axios.get(`/lists/${list_id}/${props.match.params[0]}/${id}/edit`, { headers }).catch(failure),
        axios.get(`/lists/${list_id}/users_lists`, { headers }).catch(failure),
        axios.get(`/lists/${list_id}`, { headers }).catch(failure),
      ]).then(([editListResponse, usersListsResponse, listResponse]) => {
        setUserInfo(editListResponse.headers);
        const {
          data: { item, list },
        } = editListResponse;
        const {
          data: { accepted, pending, current_user_id: currentUserId },
        } = usersListsResponse;
        const {
          data: { categories },
        } = listResponse;
        const dueByDate = formatDueBy(item.due_by);
        const acceptedUsers = accepted.map(({ user }) => user);
        const pendingUsers = pending.map(({ user }) => user);
        const currentListUsers = acceptedUsers.concat(pendingUsers);
        const userInAccepted = accepted.find(acceptedList => acceptedList.user.id === currentUserId);
        if (userInAccepted && userInAccepted.users_list.permissions === 'write') {
          setListUsers(currentListUsers);
        } else {
          // TODO: how do we pass errors around?
          props.history.push('/lists');
        }
        setUserId(item.user_id);
        setListId(list.id);
        setListType(list.type);
        setItemId(item.id);
        setProduct(item.product);
        setTask(item.task);
        setPurchased(item.purchased);
        setQuantity(item.quantity);
        setCompleted(item.completed);
        setAuthor(item.author);
        setTitle(item.title);
        setRead(item.read);
        setArtist(item.artist);
        setDueBy(dueByDate);
        setAssigneeId(item.assignee_id ? String(item.assignee_id) : '');
        setAlbum(item.album);
        setNumberInSeries(Number(item.number_in_series));
        setCategory(item.category || '');
        setCategories(categories);
      });
    }
  }, [props.history, props.match]);

  const handleSubmit = async event => {
    event.preventDefault();
    setErrors('');
    const listItem = {
      user_id: userId,
      product,
      task,
      quantity,
      purchased,
      completed,
      author,
      title,
      read,
      artist,
      album,
      due_by: dueBy,
      assignee_id: assigneeId,
      number_in_series: numberInSeries,
      category: category.trim().toLowerCase(),
    };
    listItem[`${listTypeToSnakeCase(listType)}_id`] = listId;
    const putData = {};
    putData[`${listTypeToSnakeCase(listType)}_item`] = listItem;
    try {
      const { headers } = await axios.put(
        `/lists/${listId}/${listTypeToSnakeCase(listType)}_items/${itemId}`,
        putData,
        {
          headers: JSON.parse(sessionStorage.getItem('user')),
        },
      );
      setUserInfo(headers);
      // TODO: pass success message
      props.history.push(`/lists/${listId}`);
    } catch ({ response, request, message }) {
      if (response) {
        setUserInfo(response.headers);
        if (response.status === 401) {
          // TODO: how do we pass error messages along?
          props.history.push('/users/sign_in');
        } else if (response.status === 403) {
          // TODO: how do we pass error messages along?
          props.history.push(`/lists/${listId}`);
        } else {
          const keys = Object.keys(response.data);
          const responseErrors = keys.map(key => `${key} ${response.data[key]}`);
          let joinString;
          if (listType === 'BookList' || listType === 'MusicList') {
            joinString = ' or ';
          } else {
            joinString = ' and ';
          }
          setErrors(responseErrors.join(joinString));
        }
      } else if (request) {
        // TODO: what do here?
      } else {
        setErrors(message);
      }
    }
  };

  const itemName = () =>
    ({
      BookList: `${title ? `"${title}"` : ''} ${author}`,
      GroceryList: product,
      MusicList: `${title ? `"${title}"` : ''} ${artist} ${artist && album ? `- ${album || ''}` : ''}`,
      ToDoList: task,
    }[listType]);

  const formFields = () => {
    if (listType === 'BookList') {
      return (
        <BookListItemFormFields
          author={author}
          authorChangeHandler={({ target: { value } }) => setAuthor(value)}
          title={title}
          titleChangeHandler={({ target: { value } }) => setTitle(value)}
          purchased={purchased}
          purchasedChangeHandler={() => setPurchased(!purchased)}
          read={read}
          readChangeHandler={() => setRead(!read)}
          numberInSeries={numberInSeries}
          numberInSeriesChangeHandler={({ target: { value } }) => setNumberInSeries(Number(value))}
          category={category}
          categoryChangeHandler={({ target: { value } }) => setCategory(value)}
          categories={categories}
          editForm
        />
      );
    } else if (listType === 'GroceryList') {
      return (
        <GroceryListItemFormFields
          product={product}
          productChangeHandler={({ target: { value } }) => setProduct(value)}
          quantity={quantity}
          quantityChangeHandler={({ target: { value } }) => setQuantity(value)}
          purchased={purchased}
          purchasedChangeHandler={() => setPurchased(!purchased)}
          category={category}
          categoryChangeHandler={({ target: { value } }) => setCategory(value)}
          categories={categories}
          editForm
        />
      );
    } else if (listType === 'MusicList') {
      return (
        <MusicListItemFormFields
          title={title}
          titleChangeHandler={({ target: { value } }) => setTitle(value)}
          artist={artist}
          artistChangeHandler={({ target: { value } }) => setArtist(value)}
          album={album}
          albumChangeHandler={({ target: { value } }) => setAlbum(value)}
          purchased={purchased}
          purchasedChangeHandler={() => setPurchased(!purchased)}
          category={category}
          categoryChangeHandler={({ target: { value } }) => setCategory(value)}
          categories={categories}
          editForm
        />
      );
    } else if (listType === 'ToDoList') {
      return (
        <ToDoListItemFormFields
          task={task}
          taskChangeHandler={({ target: { value } }) => setTask(value)}
          assigneeId={assigneeId}
          assigneeIdChangeHandler={({ target: { value } }) => setAssigneeId(value)}
          dueBy={dueBy}
          dueByChangeHandler={({ target: { value } }) => setDueBy(value)}
          completed={completed}
          completedChangeHandler={() => setCompleted(!completed)}
          listUsers={listUsers}
          category={category}
          categoryChangeHandler={({ target: { value } }) => setCategory(value)}
          categories={categories}
          editForm
        />
      );
    }
    return '';
  };

  return (
    <>
      <Alert errors={errors} handleDismiss={() => setErrors('')} />
      <h1>Edit {itemName()}</h1>
      <Button href={`/lists/${listId}`} className="float-right" variant="link">
        Back to list
      </Button>
      <br />
      <Form onSubmit={handleSubmit} autoComplete="off">
        {formFields()}
        <Button type="submit" variant="success" block>
          Update Item
        </Button>
      </Form>
    </>
  );
}

EditListItemForm.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string,
      list_id: PropTypes.string,
    }).isRequired,
  }).isRequired,
  history: PropTypes.shape({
    push: PropTypes.func,
  }).isRequired,
};

export default EditListItemForm;
