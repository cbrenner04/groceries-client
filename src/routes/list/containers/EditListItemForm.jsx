import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Button, Form } from 'react-bootstrap';
import { toast } from 'react-toastify';

import { listTypeToSnakeCase } from '../../../utils/format';
import axios from '../../../utils/api';
import ListItemFormFields from '../components/ListItemFormFields';
import { itemName } from '../utils';

function EditListItemForm(props) {
  const [item, setItem] = useState(props.item);

  const handleSubmit = async (event) => {
    event.preventDefault();
    const putData = {
      [`${listTypeToSnakeCase(props.list.type)}_item`]: {
        user_id: props.userId,
        product: item.product,
        task: item.task,
        quantity: item.quantity,
        purchased: item.purchased,
        completed: item.completed,
        author: item.author,
        title: item.title,
        read: item.read,
        artist: item.artist,
        album: item.album,
        due_by: item.dueBy,
        assignee_id: item.assigneeId,
        number_in_series: item.numberInSeries,
        category: item.category.trim().toLowerCase(),
        [`${listTypeToSnakeCase(props.list.type)}_id`]: props.list.id,
      },
    };
    try {
      await axios.put(
        `/lists/${props.list.id}/${listTypeToSnakeCase(props.list.type)}_items/${props.item.id}`,
        putData,
      );
      toast('Item successfully updated', { type: 'info' });
      props.history.push(`/lists/${props.list.id}`);
    } catch ({ response, request, message }) {
      if (response) {
        if (response.status === 401) {
          toast('You must sign in', { type: 'error' });
          props.history.push('/users/sign_in');
        } else if ([403, 404].includes(response.status)) {
          toast('Item not found', { type: 'error' });
          props.history.push(`/lists/${props.list.id}`);
        } else {
          const keys = Object.keys(response.data);
          const responseErrors = keys.map((key) => `${key} ${response.data[key]}`);
          let joinString;
          if (props.list.type === 'BookList' || props.list.type === 'MusicList') {
            joinString = ' or ';
          } else {
            joinString = ' and ';
          }
          toast(responseErrors.join(joinString), { type: 'error' });
        }
      } else if (request) {
        toast('Something went wrong', { type: 'error' });
      } else {
        toast(message, { type: 'error' });
      }
    }
  };

  return (
    <>
      <h1>Edit {itemName(item, props.list.type)}</h1>
      <br />
      <Form onSubmit={handleSubmit} autoComplete="off">
        <ListItemFormFields
          formData={item}
          setFormData={setItem}
          categories={props.list.categories}
          listType={props.list.type}
          listUsers={props.listUsers}
          editForm
        />
        <Button type="submit" variant="success" block>
          Update Item
        </Button>
        <Button href={`/lists/${props.list.id}`} variant="link" block>
          Cancel
        </Button>
      </Form>
    </>
  );
}

EditListItemForm.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
  listUsers: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      email: PropTypes.string.isRequired,
    }).isRequired,
  ).isRequired,
  // fetch in parent component sets default values for these properties
  item: PropTypes.shape({
    id: PropTypes.number.isRequired,
    product: PropTypes.string.isRequired,
    task: PropTypes.string.isRequired,
    purchased: PropTypes.bool.isRequired,
    quantity: PropTypes.string.isRequired,
    completed: PropTypes.bool.isRequired,
    author: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    read: PropTypes.bool.isRequired,
    artist: PropTypes.string.isRequired,
    dueBy: PropTypes.string.isRequired,
    assigneeId: PropTypes.string.isRequired,
    album: PropTypes.string.isRequired,
    numberInSeries: PropTypes.number.isRequired,
    category: PropTypes.string.isRequired,
  }).isRequired,
  list: PropTypes.shape({
    id: PropTypes.number.isRequired,
    type: PropTypes.string.isRequired,
    categories: PropTypes.arrayOf(PropTypes.string).isRequired,
  }).isRequired,
  userId: PropTypes.number.isRequired,
};

export default EditListItemForm;
