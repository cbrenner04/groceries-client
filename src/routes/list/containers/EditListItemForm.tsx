import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Form } from 'react-bootstrap';
import { toast } from 'react-toastify';
import update from 'immutability-helper';

import { list, listItem, listUsers } from '../../../types';
import axios from '../../../utils/api';
import ListItemFormFields from '../components/ListItemFormFields';
import { itemName } from '../utils';
import FormSubmission from '../../../components/FormSubmission';

function EditListItemForm(props) {
  const [item, setItem] = useState(props.item);
  const setData = ({ target: { name, value } }) => {
    let newValue = value;
    /* istanbul ignore else */
    if (name === 'numberInSeries') {
      newValue = Number(value);
    }
    const data = update(item, { [name]: { $set: newValue } });
    setItem(data);
  };

  const handleSubmit = async (event) => {
    event.preventDefault();
    const putData = {
      list_item: {
        user_id: props.userId,
        product: item.product,
        task: item.task,
        content: item.content,
        quantity: item.quantity,
        purchased: item.purchased,
        completed: item.completed,
        author: item.author,
        title: item.title,
        read: item.read,
        artist: item.artist,
        album: item.album,
        due_by: item.dueBy,
        assignee_id: item.assigneeId,
        number_in_series: item.numberInSeries,
        category: item.category.trim().toLowerCase(),
        list_id: props.list.id,
      },
    };
    try {
      await axios.put(`/lists/${props.list.id}/list_items/${props.item.id}`, putData);
      toast('Item successfully updated', { type: 'info' });
      props.navigate(`/lists/${props.list.id}`);
    } catch ({ response, request, message }) {
      if (response) {
        if (response.status === 401) {
          toast('You must sign in', { type: 'error' });
          props.navigate('/users/sign_in');
        } else if ([403, 404].includes(response.status)) {
          toast('Item not found', { type: 'error' });
          props.navigate(`/lists/${props.list.id}`);
        } else {
          const keys = Object.keys(response.data);
          const responseErrors = keys.map((key) => `${key} ${response.data[key]}`);
          let joinString;
          if (props.list.type === 'BookList' || props.list.type === 'MusicList') {
            joinString = ' or ';
          } else {
            joinString = ' and ';
          }
          toast(responseErrors.join(joinString), { type: 'error' });
        }
      } else if (request) {
        toast('Something went wrong', { type: 'error' });
      } else {
        toast(message, { type: 'error' });
      }
    }
  };

  return (
    <>
      <h1>Edit {itemName(item, props.list.type)}</h1>
      <br />
      <Form onSubmit={handleSubmit} autoComplete="off">
        <ListItemFormFields
          formData={item}
          setFormData={setData}
          categories={props.list.categories}
          listType={props.list.type}
          listUsers={props.listUsers}
          editForm
        />
        <FormSubmission
          submitText="Update Item"
          cancelAction={() => props.navigate(`/lists/${props.list.id}`)}
          cancelText="Cancel"
          displayCancelButton={true}
        />
      </Form>
    </>
  );
}

EditListItemForm.propTypes = {
  navigate: PropTypes.func.isRequired,
  listUsers: PropTypes.arrayOf(listUsers).isRequired,
  item: listItem.isRequired,
  list: list.isRequired,
  userId: PropTypes.string.isRequired,
};

export default EditListItemForm;
