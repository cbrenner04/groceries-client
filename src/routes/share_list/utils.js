import axios from '../../utils/api';
import { setUserInfo } from '../../utils/auth';

export async function fetchData({ listId, history }) {
  try {
    const { data, headers } = await axios.get(`/lists/${listId}/users_lists`, {
      headers: JSON.parse(sessionStorage.getItem('user')),
    });
    setUserInfo(headers);
    const userInAccepted = data.accepted.find(acceptedList => acceptedList.user.id === data.current_user_id);
    if (!userInAccepted || !userInAccepted.users_list.permissions === 'write') {
      history.push('/lists');
      return;
    }
    return {
      name: data.list.name,
      invitableUsers: data.invitable_users,
      listId: data.list.id,
      userIsOwner: data.user_is_owner,
      pending: data.pending,
      accepted: data.accepted,
      refused: data.refused,
      userId: data.current_user_id,
    };
  } catch ({ response, request, message }) {
    const newError = new Error();
    if (response) {
      setUserInfo(response.headers);
      if (response.status === 401) {
        // TODO: how do we pass error messages along?
        history.push('/users/sign_in');
      } else if (response.status === 403) {
        // TODO: how do we pass error messages along
        history.push('/lists');
      } else {
        const responseTextKeys = Object.keys(response.data);
        const responseErrors = responseTextKeys.map(key => `${key} ${response.data[key]}`);
        newError.message = responseErrors.join(' and ');
      }
    } else if (request) {
      newError.message = 'Something went wrong!';
    } else {
      newError.message = message;
    }
    throw newError;
  }
}
