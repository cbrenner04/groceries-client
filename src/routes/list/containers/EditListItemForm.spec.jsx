import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import { toast } from 'react-toastify';

import EditListItemForm from './EditListItemForm';
import { defaultDueBy } from '../../../utils/format';
import axios from '../../../utils/api';

jest.mock('react-toastify', () => ({
  toast: jest.fn(),
}));

describe('EditListItemForm', () => {
  let props;

  beforeEach(() => {
    props = {
      history: {
        push: jest.fn(),
      },
      listUsers: [
        {
          id: 1,
          email: 'foo@example.com',
        },
      ],
      item: {
        id: 1,
        product: 'foo',
        task: '',
        content: '',
        purchased: false,
        quantity: 'foo',
        completed: false,
        author: '',
        title: '',
        read: false,
        artist: '',
        dueBy: defaultDueBy(),
        assigneeId: '',
        album: '',
        numberInSeries: 0,
        category: '',
      },
      list: {
        id: 1,
        type: 'GroceryList',
        categories: [''],
      },
      userId: 1,
    };
  });

  it('renders', () => {
    const { container } = render(<EditListItemForm {...props} />);

    expect(container).toMatchSnapshot();
  });

  it('displays toast and redirects to list on successful submission', async () => {
    axios.put = jest.fn().mockResolvedValue({ data: {} });
    const { getByRole } = render(<EditListItemForm {...props} />);

    fireEvent.click(getByRole('button'));
    await waitFor(() => expect(axios.put).toHaveBeenCalledTimes(1));

    expect(toast).toHaveBeenCalledWith('Item successfully updated', { type: 'info' });
    expect(props.history.push).toHaveBeenCalledWith(`/lists/${props.list.id}`);
  });

  it('displays toast and redirects to login on 401', async () => {
    axios.put = jest.fn().mockRejectedValue({ response: { status: 401 } });
    const { getByRole } = render(<EditListItemForm {...props} />);

    fireEvent.click(getByRole('button'));
    await waitFor(() => expect(axios.put).toHaveBeenCalledTimes(1));

    expect(toast).toHaveBeenCalledWith('You must sign in', { type: 'error' });
    expect(props.history.push).toHaveBeenCalledWith('/users/sign_in');
  });

  it('displays toast and redirects to list on 403', async () => {
    axios.put = jest.fn().mockRejectedValue({ response: { status: 403 } });
    const { getByRole } = render(<EditListItemForm {...props} />);

    fireEvent.click(getByRole('button'));
    await waitFor(() => expect(axios.put).toHaveBeenCalledTimes(1));

    expect(toast).toHaveBeenCalledWith('Item not found', { type: 'error' });
    expect(props.history.push).toHaveBeenCalledWith(`/lists/${props.list.id}`);
  });

  it('displays toast and redirects to list on 403', async () => {
    axios.put = jest.fn().mockRejectedValue({ response: { status: 404 } });
    const { getByRole } = render(<EditListItemForm {...props} />);

    fireEvent.click(getByRole('button'));
    await waitFor(() => expect(axios.put).toHaveBeenCalledTimes(1));

    expect(toast).toHaveBeenCalledWith('Item not found', { type: 'error' });
    expect(props.history.push).toHaveBeenCalledWith(`/lists/${props.list.id}`);
  });

  it('displays appropriate error message when  and listType is BookList', async () => {
    axios.put = jest.fn().mockRejectedValue({
      response: {
        status: 500,
        data: {
          foo: 'bar',
          baz: 'foobar',
        },
      },
    });
    props.list.type = 'BookList';
    const { getByRole } = render(<EditListItemForm {...props} />);

    fireEvent.click(getByRole('button'));
    await waitFor(() => expect(axios.put).toHaveBeenCalledTimes(1));

    expect(toast).toHaveBeenCalledWith('foo bar or baz foobar', { type: 'error' });
  });

  it('displays appropriate error message when  and listType is GroceryList', async () => {
    axios.put = jest.fn().mockRejectedValue({
      response: {
        status: 500,
        data: {
          foo: 'bar',
          baz: 'foobar',
        },
      },
    });
    props.list.type = 'GroceryList';
    const { getByRole } = render(<EditListItemForm {...props} />);

    fireEvent.click(getByRole('button'));
    await waitFor(() => expect(axios.put).toHaveBeenCalledTimes(1));

    expect(toast).toHaveBeenCalledWith('foo bar and baz foobar', { type: 'error' });
  });

  it('displays appropriate error message when  and listType is MusicList', async () => {
    axios.put = jest.fn().mockRejectedValue({
      response: {
        status: 500,
        data: {
          foo: 'bar',
          baz: 'foobar',
        },
      },
    });
    props.list.type = 'MusicList';
    const { getByRole } = render(<EditListItemForm {...props} />);

    fireEvent.click(getByRole('button'));
    await waitFor(() => expect(axios.put).toHaveBeenCalledTimes(1));

    expect(toast).toHaveBeenCalledWith('foo bar or baz foobar', { type: 'error' });
  });

  it('displays appropriate error message when  and listType is ToDoList', async () => {
    axios.put = jest.fn().mockRejectedValue({
      response: {
        status: 500,
        data: {
          foo: 'bar',
          baz: 'foobar',
        },
      },
    });
    props.list.type = 'ToDoList';
    const { getByRole } = render(<EditListItemForm {...props} />);

    fireEvent.click(getByRole('button'));
    await waitFor(() => expect(axios.put).toHaveBeenCalledTimes(1));

    expect(toast).toHaveBeenCalledWith('foo bar and baz foobar', { type: 'error' });
  });

  it('displays toast when error in sending request', async () => {
    axios.put = jest.fn().mockRejectedValue({
      request: 'request failed',
    });
    props.list.type = 'ToDoList';
    const { getByRole } = render(<EditListItemForm {...props} />);

    fireEvent.click(getByRole('button'));
    await waitFor(() => expect(axios.put).toHaveBeenCalledTimes(1));

    expect(toast).toHaveBeenCalledWith('Something went wrong', { type: 'error' });
  });

  it('displays toast when unknown error', async () => {
    axios.put = jest.fn().mockRejectedValue({
      message: 'request failed',
    });
    props.list.type = 'ToDoList';
    const { getByRole } = render(<EditListItemForm {...props} />);

    fireEvent.click(getByRole('button'));
    await waitFor(() => expect(axios.put).toHaveBeenCalledTimes(1));

    expect(toast).toHaveBeenCalledWith('request failed', { type: 'error' });
  });

  it('sets value for input', async () => {
    const { getByLabelText, getByRole } = render(<EditListItemForm {...props} />);

    fireEvent.change(getByLabelText('Product'), { target: { value: 'foo' } });
    fireEvent.click(getByRole('button'));
    await waitFor(() => expect(axios.put).toHaveBeenCalledTimes(1));

    expect(axios.put).toHaveBeenCalledWith('/lists/1/grocery_list_items/1', {
      grocery_list_item: expect.objectContaining({ product: 'foo' }),
    });
  });

  it('sets value for numberInSeries as a number when input', async () => {
    props.list.type = 'BookList';
    const { getByLabelText, getByRole } = render(<EditListItemForm {...props} />);

    fireEvent.change(getByLabelText('Number in series'), { target: { value: '2' } });
    fireEvent.click(getByRole('button'));
    await waitFor(() => expect(axios.put).toHaveBeenCalledTimes(1));

    expect(axios.put).toHaveBeenCalledWith('/lists/1/book_list_items/1', {
      book_list_item: expect.objectContaining({ number_in_series: 2 }),
    });
  });
});
