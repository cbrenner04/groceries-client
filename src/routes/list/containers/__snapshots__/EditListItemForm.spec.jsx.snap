// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditListItemForm renders 1`] = `
<div>
  <h1>
    Edit 
    foo foo
  </h1>
  <br />
  <form
    autocomplete="off"
    class=""
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="product"
      >
        Product
      </label>
      <input
        class="form-control"
        id="product"
        name="product"
        placeholder="apples"
        type="text"
        value="foo"
      />
    </div>
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="quantity"
      >
        Quantity
      </label>
      <input
        class="form-control"
        id="quantity"
        name="quantity"
        placeholder="3 bags"
        type="text"
        value="foo"
      />
    </div>
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="category"
      >
        Category
      </label>
      <input
        class="form-control"
        id="category"
        list="categories"
        name="category"
        type="text"
        value=""
      />
      <datalist
        id="categories"
      >
        <option
          value=""
        />
      </datalist>
    </div>
    <div
      class="mb-3 form-group"
    >
      <div
        class="form-check"
      >
        <input
          class="form-check-input"
          id="purchased"
          name="purchased"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="purchased"
          title=""
        >
          Purchased
        </label>
      </div>
    </div>
    <button
      class="btn btn-success btn-block"
      type="submit"
    >
      Update Item
    </button>
    <a
      class="btn btn-link btn-block"
      href="/lists/1"
    >
      Cancel
    </a>
  </form>
</div>
`;
